### PR TITLE
To draw bar rois and box rois 

### PR DESCRIPTION
**Pull request by sameera2004**

### CODE REVIEW

**Comment by ericdill** [comment]
Can you provide a jupyter notebook showing the new feature in action?

**Comment by sameera2004** [comment]
Updated the ROI notebook in skbeuam-examples
https://github.com/scikit-beam/scikit-beam-examples/pull/57

File: skbeam/core/roi.py [hunk 1]
```diff
@@ -96,21 +96,23 @@ def rectangles(coords, shape):
...
-        point in image where r=0; may be a float giving subpixel precision.
-        Order is (rr, cc).
-
+    center : tuple or array
+        1. tuple
+           point in image where r=0; may be a float giving subpixel precision.
+           Order is (rr, cc).
+        2. array
+           image pixels co-ordinates
+           shape image shape
```
Comment (ericdill) [inline skbeam/core/roi.py]: I do not understand the distinction you are making between `tuple` and `array`. Is the expected value different if you pass in a `tuple` versus an `array`?

**Comment by ericdill** [comment]
Am I understanding correctly that it looks like you were making an initial attempt to extend the functionality of `rings` to include the ability to draw bars?  If so, it looks like you ended up writing a separate `bar_rois` function.  I agree that this is the correct approach.  

Can you:

- undo the changes that were made to `rings`
- rename `bar_rois` to `bars`

File: skbeam/core/roi.py (hunk 1 above)
Comment (ericdill) [inline skbeam/core/roi.py]: Deleting this comment as it is superseded by another comment

File: skbeam/core/roi.py [hunk 2]
```diff
@@ -511,3 +518,76 @@ def extract_label_indices(labels):
     label_mask = labels[labels > 0]
 
     return label_mask, pixel_list
+
+
+def bar_rois(edges, values):
```
Comment (ericdill) [inline skbeam/core/roi.py]: rename to `box`. Don't need the `_rois`, as this is in the `rois.py` module

Comment (ericdill) [inline skbeam/core/roi.py]: Err, rename to `bar`, sorry

File: skbeam/core/roi.py [hunk 3]
```diff
@@ -511,3 +518,76 @@ def extract_label_indices(labels):
...
+        specified in edges.
+
+    Note
+    ----
+    These roi's can be created using the rings function
+    """
+    return rings(edges, values, values.shape)
+
+
+def box_rois(v_values, v_edges, h_values=None, h_edges=None):
```
Comment (ericdill) [inline skbeam/core/roi.py]: Rename to `box`, same reason as above

**Comment by ericdill** [comment]
I do not like the api for `bar_rois` because it is complicated to set up and the set up code is not intuitive.  I would rather the API for this be changed from:

```python
h_values = np.zeros((20, 20))
v_values = np.zeros((20, 20))
for i in range(20):
    h_values[i, :] = i
    v_values[:, i] = i
    
edges = [[3, 4], [5, 7], [12, 15]]

h_label_array = roi.bar_rois(edges, h_values)
v_label_array = roi.bar_rois(edges, v_values)
```

to

```python
edges = [[3, 4], [5, 7], [12, 15]]
image_shape = (20,20)
h_label_array = roi.bar(edges, image_shape, horizontal=True)
h_label_array = roi.bar(edges, image_shape, horizontal=False)
```

So the function would become:

```python
def bar(edges, image_shape, horizontal=True):
    """Draw bars defined by `edges` from one edge to the other of `image_shape`

    Bars will be horizontal or vertical depending on the value of `horizontal`
    Parameters
    ----------
    edges : list
        List of tuples of inner (left or top) and outer (right or bottom)
        edges of each bar.
        e.g., edges=[(1, 2), (11, 12), (21, 22)]
    image_shape : iterable
        Shape of the image in which to create the ROIs
        e.g., image_shape=(512, 512)
    horizontal : bool, optional
        True: Make horizontal bars
        False: Make vertical bars
        Defaults to True

    Returns
    -------
    label_array : array
        Elements not inside any ROI are zero; elements inside each
        ROI are 1, 2, 3, corresponding to the order they are
        specified in `edges`.
        Has shape=`image_shape`
    """
```

**Comment by ericdill** [comment]
Please disregard my previous comment about undoing the changes to `rings`. Instead of undoing the changes to rings, I would advocate for refactoring `rings` so that the commonality between `rings` and `bar` can be shared, so it looks more like this:

```python
def _make_roi(coords, edges, shape):
    label_array = np.digitize(coords, edges, right=False)
    # Even elements of label_array are in the space between rings.
    label_array = (np.where(label_array % 2 != 0, label_array, 0) + 1) // 2
    return label_array.reshape(shape)


def rings(edges, center, shape):
    edges = np.atleast_2d(np.asarray(edges)).ravel()
    r_coord = utils.radial_grid(center, shape).ravel()
    return _make_roi(r_coord, edges)


def bar(edges, shape, horizontal=True):
    # function setup
    values = np.repeat(range(shape[0]), shape[1])
    if not horizontal:
        values = values.reshape(shape).T.ravel()
    return _make_roi(values, edges, shape)
```

I stripped the docstrings and error handling so that my proposed change is more clear.  (Obviously keep the docstrings and input validation :smile: )

Also note that I have not tested this code, so treat it as basically pseudo-code...

Thanks @sameera2004 , this is a very valuable addition to the `roi` module!

File: skbeam/core/roi.py (hunk 3 above)
Comment (ericdill) [inline skbeam/core/roi.py]: I would change this API to be:

```def box(v_edges, h_edges, shape)```

The set up for `v_values` and `h_values` is not intuitive and we can do that inside this function

File: skbeam/core/tests/test_roi.py [hunk 4]
```diff
@@ -341,3 +341,27 @@ def test_kymograph():
...
+    box_array = roi.box_rois(v_values, edges, h_values)
+
+    assert_array_equal(box_array[3, 3], 1)
+    assert_array_equal(box_array[5, 5], 4)
+    assert_array_equal(np.unique(box_array)[1:], np.array([1, 2, 3, 4]))
+
+    n_edges = edges = [[3, 4], [5, 7], [8]]
+    h_val, v_val = np.mgrid[0:20, 0:20]
+    assert_raises(ValueError, lambda: roi.box_rois(v_values, n_edges))
+    assert_raises(ValueError, lambda: roi.box_rois(v_values, edges, h_val))
```
Comment (ericdill) [inline skbeam/core/tests/test_roi.py]: Drop the `lambda`. `assert_raises` works like this:
```python
assert_raises(Exception, callable, *args, **kwargs)
```
Under the hood assert_raises can be understood to be a context manager that calls `callable(*args, **kwargs)` and checks that `ValueError` was raised, so you should write your `assert_raises` like this:

```python
assert_raise(ValueError, roi.box_rois, v_values, edges, h_val)
```

File: skbeam/core/roi.py (hunk 3 above)
Comment (sameera2004) [inline skbeam/core/roi.py]: Sure

**Comment by sameera2004** [comment]
Ready for review

File: skbeam/core/roi.py [hunk 5]
```diff
@@ -106,11 +106,9 @@ def rings(edges, center, shape):
     edges: list
         giving the inner and outer radius of each ring
         e.g., [(1, 2), (11, 12), (21, 22)]
-
     center : tuple
-        point in image where r=0; may be a float giving subpixel precision.
-        Order is (rr, cc).
-
+           point in image where r=0; may be a float giving subpixel precision.
```
Comment (ericdill) [inline skbeam/core/roi.py]: extra tab here

File: skbeam/core/roi.py [hunk 6]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
...
+    Note
+    ----
+    To draw boxes according to the image pixels co-ordinates has to provide
+    both h_values and v_values
+
+    """
+    if h_edges is None:
+        h_edges = v_edges
+
+    if h_values is None and v_values is None :
```
Comment (ericdill) [inline skbeam/core/roi.py]: the extra space before the colon here is why travis is failing

File: skbeam/core/roi.py [hunk 7]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
...
+    # Even elements of label_array are in the space between rings.
+    label_array = (np.where(label_array % 2 != 0, label_array, 0) + 1) // 2
+    return label_array.reshape(shape)
+
+
+def bar(edges, shape, horizontal=True, values=None):
+    """Draw bars defined by `edges` from one edge to the other of `image_shape`
+
+    Bars will be horizontal or vertical depending on the value of `horizontal`
+    Parameters
```
Comment (ericdill) [inline skbeam/core/roi.py]: Add a space between this line and the one above

File: skbeam/core/roi.py [hunk 8]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
...
+    shape : tuple
+        Shape of the image in which to create the ROIs
+        e.g., shape=(512, 512)
+    horizontal : bool, optional
+        True: Make horizontal bars
+        False: Make vertical bars
+        Defaults to True
+    values : array, optional
+        image pixels co-ordinates
+    Returns
```
Comment (ericdill) [inline skbeam/core/roi.py]: Add a space between this line and the one above

File: skbeam/core/roi.py [hunk 9]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
...
+    h_edges : list, optional
+        giving the inner and outer edges of each horizontal bar
+        e.g., [(1, 2), (11, 12), (21, 22)]
+    h_values : array, optional
+        image pixels co-ordinates in horizontal direction
+        shape has to be image shape
+    v_values : array, optional
+        image pixels co-ordinates in vertic
+        shape has to be image shape
+    Returns
```
Comment (ericdill) [inline skbeam/core/roi.py]: Add a space between this line and the one above

File: skbeam/core/roi.py [hunk 10]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
...
+    coords : array
+        shape is image shape
+    edges : list
+        List of tuples of inner (left or top) and outer (right or bottom)
+        edges of each roi.
+        e.g., edges=[(1, 2), (11, 12), (21, 22)]
+    shape : tuple
+        Shape of the image in which to create the ROIs
+        e.g., shape=(512, 512)
+    """
```
Comment (ericdill) [inline skbeam/core/roi.py]: Add a `Returns` section please

**Comment by ericdill** [comment]
Can you add a note to `bar()` and `box()` that the primary use case (for now) is in GISAXS? That would probably be enough to keep others from having the same confusion I did :smile:

File: skbeam/core/roi.py [hunk 11]
```diff
@@ -511,3 +505,123 @@ def extract_label_indices(labels):
     label_mask = labels[labels > 0]
 
     return label_mask, pixel_list
+
+
+def _make_roi(coords, edges, shape):
+    """ Helper function to create ring roi's and bar roi's
```
Comment (ericdill) [inline skbeam/core/roi.py]: I'm not sure you should have an apostrophe in `roi's`

**Comment by sameera2004** [comment]
Added a note about GISXS use case

File: skbeam/core/roi.py [hunk 12]
```diff
@@ -511,3 +505,138 @@ def extract_label_indices(labels):
...
+    Note
+    ----
+    The primary use case is in GISAXS.
+    """
+    edges = np.atleast_2d(np.asarray(edges)).ravel()
+    if not 0 == len(edges) % 2:
+        raise ValueError("edges should have an even number of elements, "
+                         "giving inner, outer edge value for each bar")
+    if not np.all(np.diff(edges) >= 0):
+        raise ValueError("edges are expected to be monotonically increasing, "
```
Comment (ericdill) [inline skbeam/core/roi.py]: Can you add a test to execute this line?

File: skbeam/core/roi.py [hunk 13]
```diff
@@ -511,3 +505,138 @@ def extract_label_indices(labels):
...
+        specified in `edges`.
+        Has shape=`image shape`
+
+    Note
+    ----
+    The primary use case is in GISAXS.
+    """
+    edges = np.atleast_2d(np.asarray(edges)).ravel()
+    if not 0 == len(edges) % 2:
+        raise ValueError("edges should have an even number of elements, "
```
Comment (ericdill) [inline skbeam/core/roi.py]: Can you make sure to hit this line in the tests?

File: skbeam/core/roi.py [hunk 14]
```diff
@@ -131,15 +129,11 @@ def rings(edges, center, shape):
                          "giving inner and outer radii of each ring from "
```
Comment (ericdill) [inline skbeam/core/roi.py]: Can you make sure to hit this line in the tests? Also, line 125 is not covered

File: skbeam/core/roi.py (hunk 12 above)
Comment (sameera2004) [inline skbeam/core/roi.py]: Sure, I will add

File: skbeam/core/roi.py (hunk 13 above)
Comment (sameera2004) [inline skbeam/core/roi.py]: I already added a test for this line to find the `edges to have even number of elements`. line 125 already covered but not line 127 I am adding one now

**Comment by sameera2004** [comment]
Ready for review

**Comment by ericdill** [comment]
Great, thanks @sameera2004 , can you please make sure to update the notebook?

**Comment by sameera2004** [comment]
Thanks, @ericdill

**Pull request merged**